File: desktop/src/db/account-refresh-repository.ts
/**
 * 账号刷新任务仓储（Node 重写）
 * 对标 services/repositories/account_refresh_repository.py
 *
 * 两张表：account_refresh_tasks（任务）+ account_refresh_task_items（明细）
 *
 * 时间字段的重要差异：Python 侧用 datetime.now() 作为**参数**传入，
 * 而不是 SQL 的 CURRENT_TIMESTAMP。sqlite3 适配器会把它序列化成
 * "YYYY-MM-DD HH:MM:SS.ffffff"（本地时间，含微秒）。
 * 这里用 nowLiteral() 生成同格式字符串，保证两侧写入的数据可互读。
 *
 * 本仓储不吞异常（与 Python 一致，失败会直接抛出）。
 */
import type { Db } from "./connection.ts";

/**
 * 生成与 Python datetime.now() 等价的字面量：本地时间 + 微秒。
 * 例：2026-09-23 05:31:07.123456
 */
export function nowLiteral(d: Date = new Date()): string {
  const p = (n: number, w = 2) => String(n).padStart(w, "0");
  const micro = `${p(d.getMilliseconds(), 3)}000`;
  return (
    `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ` +
    `${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}.${micro}`
  );
}

/** 单条明细的刷新结果 */
export interface RefreshItemResult {
  error_message?: string | null;
  is_pro?: string | null;
  pro_plan_name?: string | null;
  family_role?: string | null;
  family_manager_email?: string | null;
  has_family_group?: string | null;
  family_member_count?: number | null;
  family_slots_left?: number | null;
  account_country?: string | null;
}

export interface RefreshTaskRow {
  id: number;
  task_type: string | null;
  task_mode: string | null;
  status: string | null;
  total_count: number | null;
  progress_current: number | null;
  progress_percent: number | null;
  success_count: number | null;
  failed_count: number | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string | null;
  [key: string]: unknown;
}

export interface RefreshTaskItemRow {
  id: number;
  task_id: number;
  email: string;
  status: string | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
  [key: string]: unknown;
}

export class AccountRefreshRepository {
  private readonly db: Db;

  constructor(db: Db) {
    this.db = db;
  }

  /**
   * 创建刷新任务，返回 task_id。
   * 注意 task_type 在 Python 侧被硬编码为 "family_info_refresh"。
   */
  createRefreshTask(taskMode: string, totalCount: number): number {
    const info = this.db
      .prepare(
        `INSERT INTO account_refresh_tasks
         (task_type, task_mode, status, total_count, started_at)
         VALUES (?, ?, 'running', ?, ?)`,
      )
      .run("family_info_refresh", taskMode, totalCount, nowLiteral());
    return Number(info.lastInsertRowid);
  }

  /** 批量创建明细，初始状态 pending。Python 侧是逐条 INSERT，这里用事务包住提速。 */
  createTaskItems(taskId: number, emails: string[]): void {
    const stmt = this.db.prepare(
      `INSERT INTO account_refresh_task_items (task_id, email, status)
       VALUES (?, ?, 'pending')`,
    );
    this.db.exec("BEGIN");
    try {
      for (const email of emails) stmt.run(taskId, email);
      this.db.exec("COMMIT");
    } catch (error) {
      this.db.exec("ROLLBACK");
      throw error;
    }
  }

  /** 标记明细开始执行 */
  updateTaskItemStarted(taskId: number, email: string): void {
    this.db
      .prepare(
        `UPDATE account_refresh_task_items SET
           status = 'running',
           started_at = ?
         WHERE task_id = ? AND email = ?`,
      )
      .run(nowLiteral(), taskId, email);
  }

  /** 写入明细结果。11 个字段的绑定顺序与 Python 完全一致。 */
  updateTaskItem(taskId: number, email: string, status: string, result: RefreshItemResult): void {
    this.db
      .prepare(
        `UPDATE account_refresh_task_items SET
           status = ?,
           error_message = ?,
           is_pro = ?,
           pro_plan_name = ?,
           family_role = ?,
           family_manager_email = ?,
           has_family_group = ?,
           family_member_count = ?,
           family_slots_left = ?,
           account_country = ?,
           finished_at = ?
         WHERE task_id = ? AND email = ?`,
      )
      .run(
        status,
        result.error_message ?? null,
        result.is_pro ?? null,
        result.pro_plan_name ?? null,
        result.family_role ?? null,
        result.family_manager_email ?? null,
        result.has_family_group ?? null,
        result.family_member_count ?? null,
        result.family_slots_left ?? null,
        result.account_country ?? null,
        nowLiteral(),
        taskId,
        email,
      );
  }

  /**
   * 更新任务进度。
   * 百分比在应用侧算：total_count 为 0 时取 0，读不到任务时 total_count 兜底为 1（与 Python 一致）。
   */
  updateTaskProgress(
    taskId: number,
    progressCurrent: number,
    successCount: number,
    failedCount: number,
  ): void {
    const row = this.db
      .prepare("SELECT total_count FROM account_refresh_tasks WHERE id = ?")
      .get(taskId) as { total_count: number } | undefined;
    const totalCount = row ? row.total_count : 1;
    const progressPercent = totalCount > 0 ? (progressCurrent / totalCount) * 100 : 0;

    this.db
      .prepare(
        `UPDATE account_refresh_tasks SET
           progress_current = ?,
           progress_percent = ?,
           success_count = ?,
           failed_count = ?
         WHERE id = ?`,
      )
      .run(progressCurrent, progressPercent, successCount, failedCount, taskId);
  }

  /** 收尾任务：进度强制拉满（progress_current = total_count, percent = 100.0） */
  finishTask(taskId: number, status: string, successCount: number, failedCount: number): void {
    this.db
      .prepare(
        `UPDATE account_refresh_tasks SET
           status = ?,
           success_count = ?,
           failed_count = ?,
           progress_current = total_count,
           progress_percent = 100.0,
           finished_at = ?
         WHERE id = ?`,
      )
      .run(status, successCount, failedCount, nowLiteral(), taskId);
  }

  getTaskById(taskId: number): RefreshTaskRow | null {
    const row = this.db
      .prepare("SELECT * FROM account_refresh_tasks WHERE id = ?")
      .get(taskId);
    return (row as RefreshTaskRow) ?? null;
  }

  getTaskItems(taskId: number): RefreshTaskItemRow[] {
    return this.db
      .prepare("SELECT * FROM account_refresh_task_items WHERE task_id = ? ORDER BY id")
      .all(taskId) as RefreshTaskItemRow[];
  }

  getRecentTasks(limit = 10): RefreshTaskRow[] {
    return this.db
      .prepare("SELECT * FROM account_refresh_tasks ORDER BY created_at DESC LIMIT ?")
      .all(limit) as RefreshTaskRow[];
  }
}

File: services/repositories/account_refresh_repository.py
"""账号刷新任务仓库。

提供账号会员信息刷新任务的持久化操作。
"""

from __future__ import annotations

import threading
from datetime import datetime
from typing import Callable, List, Optional


class AccountRefreshRepository:
    """账号刷新任务数据访问层"""

    @staticmethod
    def create_refresh_task(
        task_mode: str,
        total_count: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ) -> int:
        """
        创建刷新任务，返回任务ID

        Args:
            task_mode: 任务模式 ('pro_only' | 'full')
            total_count: 总账号数
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁

        Returns:
            int: 新建任务的 ID
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT INTO account_refresh_tasks
                (task_type, task_mode, status, total_count, started_at)
                VALUES (?, ?, 'running', ?, ?)
                """,
                ("family_info_refresh", task_mode, total_count, datetime.now()),
            )
            task_id = cursor.lastrowid
            conn.commit()
            conn.close()
            return task_id

    @staticmethod
    def create_task_items(
        task_id: int,
        emails: List[str],
        connection_factory: Callable,
        db_lock: threading.Lock,
    ):
        """
        批量创建任务明细

        Args:
            task_id: 任务 ID
            emails: 邮箱列表
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            for email in emails:
                cursor.execute(
                    """
                    INSERT INTO account_refresh_task_items (task_id, email, status)
                    VALUES (?, ?, 'pending')
                    """,
                    (task_id, email),
                )
            conn.commit()
            conn.close()

    @staticmethod
    def update_task_item_started(
        task_id: int,
        email: str,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ):
        """
        标记任务明细开始执行

        Args:
            task_id: 任务 ID
            email: 账号邮箱
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE account_refresh_task_items SET
                    status = 'running',
                    started_at = ?
                WHERE task_id = ? AND email = ?
                """,
                (datetime.now(), task_id, email),
            )
            conn.commit()
            conn.close()

    @staticmethod
    def update_task_item(
        task_id: int,
        email: str,
        status: str,
        result: dict,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ):
        """
        更新单条任务明细

        Args:
            task_id: 任务 ID
            email: 账号邮箱
            status: 状态 ('success' | 'failed' | 'skipped')
            result: 刷新结果字典
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE account_refresh_task_items SET
                    status = ?,
                    error_message = ?,
                    is_pro = ?,
                    pro_plan_name = ?,
                    family_role = ?,
                    family_manager_email = ?,
                    has_family_group = ?,
                    family_member_count = ?,
                    family_slots_left = ?,
                    account_country = ?,
                    finished_at = ?
                WHERE task_id = ? AND email = ?
                """,
                (
                    status,
                    result.get("error_message"),
                    result.get("is_pro"),
                    result.get("pro_plan_name"),
                    result.get("family_role"),
                    result.get("family_manager_email"),
                    result.get("has_family_group"),
                    result.get("family_member_count"),
                    result.get("family_slots_left"),
                    result.get("account_country"),
                    datetime.now(),
                    task_id,
                    email,
                ),
            )
            conn.commit()
            conn.close()

    @staticmethod
    def update_task_progress(
        task_id: int,
        progress_current: int,
        success_count: int,
        failed_count: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ):
        """
        更新任务进度

        Args:
            task_id: 任务 ID
            progress_current: 当前进度
            success_count: 成功数
            failed_count: 失败数
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()

            # 先获取 total_count 计算百分比
            cursor.execute(
                "SELECT total_count FROM account_refresh_tasks WHERE id = ?",
                (task_id,),
            )
            row = cursor.fetchone()
            total_count = row[0] if row else 1
            progress_percent = (progress_current / total_count * 100) if total_count > 0 else 0

            cursor.execute(
                """
                UPDATE account_refresh_tasks SET
                    progress_current = ?,
                    progress_percent = ?,
                    success_count = ?,
                    failed_count = ?
                WHERE id = ?
                """,
                (progress_current, progress_percent, success_count, failed_count, task_id),
            )
            conn.commit()
            conn.close()

    @staticmethod
    def finish_task(
        task_id: int,
        status: str,
        success_count: int,
        failed_count: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ):
        """
        完成任务

        Args:
            task_id: 任务 ID
            status: 最终状态 ('completed' | 'failed' | 'stopped')
            success_count: 成功数
            failed_count: 失败数
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE account_refresh_tasks SET
                    status = ?,
                    success_count = ?,
                    failed_count = ?,
                    progress_current = total_count,
                    progress_percent = 100.0,
                    finished_at = ?
                WHERE id = ?
                """,
                (status, success_count, failed_count, datetime.now(), task_id),
            )
            conn.commit()
            conn.close()

    @staticmethod
    def get_task_by_id(
        task_id: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ) -> Optional[dict]:
        """
        获取任务详情

        Args:
            task_id: 任务 ID
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁

        Returns:
            dict | None: 任务信息字典
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                "SELECT * FROM account_refresh_tasks WHERE id = ?",
                (task_id,),
            )
            row = cursor.fetchone()
            conn.close()
            if row:
                return dict(row)
            return None

    @staticmethod
    def get_task_items(
        task_id: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ) -> List[dict]:
        """
        获取任务明细列表

        Args:
            task_id: 任务 ID
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁

        Returns:
            List[dict]: 任务明细列表
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                "SELECT * FROM account_refresh_task_items WHERE task_id = ? ORDER BY id",
                (task_id,),
            )
            rows = cursor.fetchall()
            conn.close()
            return [dict(row) for row in rows]

    @staticmethod
    def get_recent_tasks(
        limit: int,
        connection_factory: Callable,
        db_lock: threading.Lock,
    ) -> List[dict]:
        """
        获取最近的任务列表

        Args:
            limit: 返回数量限制
            connection_factory: 数据库连接工厂
            db_lock: 数据库锁

        Returns:
            List[dict]: 任务列表
        """
        with db_lock:
            conn = connection_factory()
            cursor = conn.cursor()
            cursor.execute(
                """
                SELECT * FROM account_refresh_tasks
                ORDER BY created_at DESC
                LIMIT ?
                """,
                (limit,),
            )
            rows = cursor.fetchall()
            conn.close()
            return [dict(row) for row in rows]
